### ISSUE TITLE
The "compiler" example doesn't actually show how to use `Interner` realistically

### ISSUE THREAD

**Issue by regexident** [salsa-rs/salsa#151]
The ["compiler" example](https://github.com/salsa-rs/salsa/blob/master/examples/compiler/implementation.rs#L21-L25) includes an `interner: Interner`, which is meant to illustrate the use of shared mutable state:

```rust
#[salsa::database(class_table::ClassTable)]
#[derive(Default)]
pub struct DatabaseImpl {
    runtime: salsa::Runtime<DatabaseImpl>,

    /// An interner is an example of shared mutable state that would
    /// be ok: although the interner allocates internally when you
    /// intern something new, this never affects any previously
    /// interned values, so it's not going to affect query results.
    interner: Interner,
}
```

The example however doesn't make use of `interner` at all in its illustrated API/functionality.

As such `fn interner(&self) -> &Interner;` only gives non-mutable access to `interner`.

When I changed it to `fn interner(&mut self) -> &mut Interner;` and added a corresponding method to the query group:

```rust
pub struct Dummy;

pub trait DummyDatabase: salsa::Database {
    fn interner(&mut self) -> &mut Interner;
}

impl Interner {
    fn intern(&mut self, s: &Dummy) -> u32 {
        unimplemented!()
    }
}

#[salsa::query_group(Outputs)]
pub trait OutputsDatabase: MachineDatabase {
    fn dummy(&mut self, dummy: Dummy) -> Arc<u32>;
}
```

I'm getting this error:

```
custom attribute panicked

help: message: first argument of query `dummy` must be `&self` or `&mut self`
```

which is confusing as the query's first argument is `&mut self` afaict.

Long story short, it would be nice to adapt the "compiler" example to be a slightly more realistic scenario. Preferrably even showing how interning would actually be used with salsa (as in actually contributing to solving an example problem). From the example alone it's not clear why one would want to intern stuff to begin with.

**Comment by bjorn3** [comment]
> As such fn interner(&self) -> &Interner; only gives non-mutable access to interner.

An interner should probably use inner mutability. Eg wrap the state in a `RefCell` or `Mutex`.

**Comment by matklad** [comment]
The specific case of interning wold be handled in a first-class manner by https://github.com/salsa-rs/salsa/pull/150. An example of realistic salsa-compatible interner could be found in rust-analyzer: https://github.com/rust-analyzer/rust-analyzer/blob/d080c8f02105bc52f069785ae3e843a6606560e1/crates/ra_db/src/loc2id.rs

**Comment by frondeus** [comment]
Still the error message is wrong. It says "or `&mut self`" but inside macro we check if mutability is none.

**Comment by regexident** [comment]
With https://github.com/salsa-rs/salsa/pull/150 merged could we update the "compiler" example to give a more realistic (i.e. non-contrived) picture of how interning would be used with salsa? The current example declares an `Interner` type but keeps it unimplemented dead code, execution-wise.

It also doesn't show how interned values would be of use with/within queries.

I know what interning is, how it's implemented and why it's used in compilers but am still a bit lost in the context of salsa.

**Comment by regexident** [comment]
@heyimalex even having understood the purpose and possibly even the implementation of interning and having read the corresponding Salsa RFC there's still quite a gap between the abstract concept and an actual example project using the API with code that actually does stuff that might resemble what you might be doing in your own scenario.

I just gave it a shot myself: https://github.com/salsa-rs/salsa/pull/202
